app/src/main/java/com/sschuraytz/charcoaldrawing/MainActivity.java

```diff
@@ -32,11 +34,13 @@ protected void onCreate(Bundle savedInstanceState) {
         setUpRedo();
     }
 
+    @SuppressLint("ClickableViewAccessibility")
     public void setUpDrawingView()
     {
         drawingView = (DrawingView) findViewById(R.id.canvas);
         drawingView.setOnTouchListener((v, event) -> {
-            updateUndoVisibility();
+            updateVisibility(drawingView.undoRedo.getCurrentStackSize() - 1, undoButton);
+            updateVisibility(drawingView.undoRedo.getUndoneStackSize(), redoButton);
```

Comment:
In order to make a method general enough to abstract my code, I had to assume all passed in stacks will have a same minimum size. But, I didn't create my new stack for "redo" with an ever-present 0th bitmap (like we did for "undo".) Thus, I have to add the " - 1" to the parameter for undo. Is this messy?

Reply:
This wasn't exactly how I meant you to do it. 
What I want you to do is create an interface called something like UndoRedoListener with two methods, onUndoAvailable(boolean) and onRedoAvailable(boolean).

You'll add a setListener() method to UndoRedo and then when undo() or redo() is called it will call the method in the listener passing true or false if the feature is available.

Then you'll implement an UndoRedoListener in your Activity.